common/src/main/java/com/wynntils/gui/render/MapRenderer.java

```diff
@@ -99,12 +109,137 @@ public static void renderMapQuad(
+
+                    return poi instanceof ServicePoi && inMap;
+                })
+                .map(poi -> (ServicePoi) poi)
+                .toList();
+
+        renderServicePois(
+                poseStack,
+                mouseX,
+                mouseY,
+                mapCenterX,
+                mapCenterZ,
+                centerX,
+                centerZ,
+                scale,
+                poiScale,
+                followPlayerRotation,
+                servicePois);
+
         // disable rotation if necessary
         if (followPlayerRotation) {
             poseStack.popPose();
         }
     }
 
+    private static void renderServicePois(
+            PoseStack poseStack,
+            int mouseX,
+            int mouseY,
+            float mapCenterX,
+            float mapCenterZ,
+            float centerX,
+            float centerZ,
+            float scale,
+            float poiScale,
+            boolean followPlayerRotation,
+            List<ServicePoi> servicePois) {
+        poseStack.pushPose();
+
+        if (followPlayerRotation) {
+            poseStack.mulPose(new Quaternion(
+                    0F,
+                    0,
+                    (float) StrictMath.sin(Math.toRadians(180 + McUtils.player().getYRot()) / 2),
+                    (float) StrictMath.cos(
+                            -Math.toRadians(180 + McUtils.player().getYRot()) / 2)));
+        }
+
+        for (ServicePoi servicePoi : servicePois) {
+            renderServicePoi(
+                    poseStack, mouseX, mouseY, mapCenterX, mapCenterZ, centerX, centerZ, scale, poiScale, servicePoi);
+        }
+
+        if (followPlayerRotation) {
+            poseStack.popPose();
+        }
+    }
+
+    public static void renderServicePoi(
+            PoseStack poseStack,
+            int mouseX,
+            int mouseY,
+            float mapCenterX,
+            float mapCenterZ,
+            float centerX,
+            float centerZ,
+            float scale,
+            float poiScale,
+            ServicePoi servicePoi) {
+        // TODO: This is really basic at the moment
+        //       Add fading, and other configs
+
+        double distanceX = servicePoi.getLocation().getX() - mapCenterX;
+        double distanceZ = servicePoi.getLocation().getZ() - mapCenterZ;
+
+        float textureXPosition = (float) (centerX + distanceX / scale);
+        float textureZPosition = (float) (centerZ + distanceZ / scale);
+
+        float width = servicePoi.getIcon().width() * poiScale;
+        float height = servicePoi.getIcon().height() * poiScale;
+
+        poseStack.pushPose();
+
+        float renderX = textureXPosition - width / 2f;
+        float renderZ = textureZPosition - height / 2f;
+        poseStack.translate(renderX, renderZ, 0);
+
+        if (mouseX >= renderX && mouseX <= renderX + width && mouseY >= renderZ && mouseY <= renderZ + height) {
+            FontRenderer.getInstance()
+                    .renderText(
```

Comment:
I think it's great if we have the possibility to have text labels for all POIs, but I also think this needs to be a setting which should probably be off by default (and in the future, perhaps setting it to on/off differently depending on what kind of POI it is, and zoom level etc).

Reply:
This only works on main map, on hover. I think it is fine to leave it on.

Reply:
Aha, it is on hover. I missed that part. That is very good!